### PR TITLE
Add support for Unix sockets #1118

### PR DESCRIPTION
**Pull request by graingert**
This also adds support for TLS!

Simply use https://exmaple.com/ for TLS or [http+unix://%2Fvar%2Frun%2Fluigid%2Fluigid.sock](http:/%2Fvar%2Frun%2Fluigid%2Fluigid.sock) for Unix sockets

There doesn't seem to be any actual tests to see if the real fetch method works, as all the tests that cover RemoteScheduler patches "._fetch"

### CODE REVIEW

**Comment by Tarrasch** [comment]
Hmm, can you elaborate more on what this is and who would use it and why? Is this only related to the luigi server?

**Comment by graingert** [comment]
@Tarrasch the discussion is in #1118

**Comment by erikbern** [comment]
lgtm

File: test/worker_test.py [hunk 1]
```diff
@@ -649,7 +649,7 @@ def tearDown(self):
 
     @email_patch
     def test_connection_error(self, emails):
-        sch = RemoteScheduler(host="this_host_doesnt_exist", port=1337, connect_timeout=1)
```
Comment (graingert) [inline test/worker_test.py]: this_host_doesnt_exist is a perfectly valid option for a gTLD

File: luigi/rpc.py [hunk 2]
```diff
@@ -42,32 +62,61 @@ def __init__(self, message, sub_exception=None):
...
+class FetcherException(Exception):
+    def __init__(self, original_exc):
+        self.original_exc = original_exc
+
+
+class URLLibFetcher(object):
+    def fetch(self, full_url, body, timeout):
+        try:
+            body = urlencode(body).encode('utf-8')
+            return urlopen(full_url, body, self._connect_timeout).read().decode('utf-8')
```
Comment (Tarrasch) [inline luigi/rpc.py]: Eh this is scary. There's no such thing as a `self._connect_timeout` in this context right? Is this code path really tested?

I know I've had to disable some tests on Travis. Can you make sure a normal `tox -e py27-nonhdfs` passes before we merge this?

File: tox.ini [hunk 3]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py{27,33,34}-{cdh,hdp,nonhdfs,gcloud,postgres}, pypy-scheduler, docs, pep8
+envlist = py{27,33,34}-{cdh,hdp,nonhdfs,gcloud,postgres}{-requests,-unix,}, pypy-scheduler, docs, pep8
```
Comment (Tarrasch) [inline tox.ini]: I don't see any corresponding `attr('requests')`, nor did you add it to the `.travis.yml` right? Also i don't see why you need multiplicative expansion here. Why not just `py27-unix`? Also something more explicit over just `unix`. Like `py27-unixsockets` or something.

File: luigi/rpc.py (hunk 2 above)
Comment (graingert) [inline luigi/rpc.py]: Ah woops, looks like something depends on requests in your test dependencies. Will fix

**Comment by graingert** [comment]
@Tarrasch thanks for the review, I've made those changes.

File: tox.ini [hunk 4]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py{27,33,34}-{cdh,hdp,nonhdfs,gcloud,postgres}, pypy-scheduler, docs, pep8
+envlist = py{27,33,34}-{cdh,hdp,nonhdfs,gcloud,postgres}{-unixsockets,}, pypy-scheduler, docs, pep8
```
Comment (Tarrasch) [inline tox.ini]: Again, why multiplicative?

File: luigi/rpc.py [hunk 5]
```diff
@@ -42,32 +62,61 @@ def __init__(self, message, sub_exception=None):
...
+class FetcherException(Exception):
+    def __init__(self, original_exc):
+        self.original_exc = original_exc
+
+
+class URLLibFetcher(object):
+    def fetch(self, full_url, body, timeout):
+        try:
+            body = urlencode(body).encode('utf-8')
+            return urlopen(full_url, body, timeout).read().decode('utf-8')
```
Comment (Tarrasch) [inline luigi/rpc.py]: Awesome that you fixed this! But you checked somehow that this code path is tested before fixing it right? :)

File: luigi/rpc.py [hunk 6]
```diff
@@ -25,13 +25,33 @@
...
+    import requests_unixsockets as requests
+except ImportError:
+    HAS_UNIX_SOCKETS = False
+    try:
+        import requests
+    except ImportError:
+        HAS_REQUESTS = False
+
+try:
+    from requests import exceptions as requests_exceptions
```
Comment (Tarrasch) [inline luigi/rpc.py]: Hmm, if you don't have it installed, won't you just get an exception like "no variable defined `requests_exceptions
Markdown supported `"? Can't you just do `from requests import exceptions as requests_exceptions` without try-catch inside the `fetch()` method?

File: test/server_test.py [hunk 7]
```diff
@@ -25,8 +26,15 @@
 import luigi.rpc
 import luigi.server
 from luigi.scheduler import CentralPlannerScheduler
+from luigi.six.moves.urllib.parse import urlencode, ParseResult
 
 from tornado.testing import AsyncHTTPTestCase
+from nose.plugins.attrib import attr
+
+try:
+    from unittest import mock
```
Comment (Tarrasch) [inline test/server_test.py]: You should be able to assume that `unittest` is installed I believe.

File: test/server_test.py [hunk 8]
```diff
@@ -72,19 +81,50 @@ def test_api_404(self):
         self._test_404('/api/foo')
 
 
+class INETServerClient(object):
```
Comment (Tarrasch) [inline test/server_test.py]: What is "INET"?

**Comment by Tarrasch** [comment]
Starting to look better, but I want to make sure that this part of the code hold high quality as it's so central. :)

File: test/server_test.py (hunk 8 above)
Comment (graingert) [inline test/server_test.py]: Socket family

https://docs.python.org/2/library/socket.html#socket.AF_UNIX
https://docs.python.org/2/library/socket.html#socket.AF_INET

File: test/server_test.py (hunk 7 above)
Comment (graingert) [inline test/server_test.py]: Python 3.4 puts mock on unittest.mock

File: luigi/rpc.py (hunk 5 above)
Comment (graingert) [inline luigi/rpc.py]: https://github.com/spotify/luigi/pull/1129/files#diff-9b595b54a300214f2172cde2e9c1eec4R169

File: luigi/rpc.py (hunk 6 above)
Comment (graingert) [inline luigi/rpc.py]: yep that makes more sense

**Comment by graingert** [comment]
@Tarrasch thanks!

File: test/server_test.py (hunk 7 above)
Comment (Tarrasch) [inline test/server_test.py]: Ah ok :)

**Comment by Tarrasch** [comment]
Looks good to me now :)

File: test/server_test.py (hunk 8 above)
Comment (erikbern) [inline test/server_test.py]: internet :)

it's pretty cool, you should try it!

File: luigi/rpc.py [hunk 9]
```diff
@@ -42,32 +57,62 @@ def __init__(self, message, sub_exception=None):
...
         time.sleep(30)
 
     def _fetch(self, url, body, log_exceptions=True, attempts=3):
-
-        full_url = 'http://{host}:{port:d}{prefix}{url}'.format(
-            host=self._host,
-            port=self._port,
-            prefix=self._url_prefix,
-            url=url)
+        full_url = self._url + url
```
Comment (Tarrasch) [inline luigi/rpc.py]: Erm, there's no way this line can be "right". Are the variable names extremely confusing? Or you actually string-concatenating two urls?

Comment (graingert) [inline luigi/rpc.py]: `full_url = self._url + url_suffix`?

Comment (Tarrasch) [inline luigi/rpc.py]: Yea, that's *much* clearer.

**Comment by Tarrasch** [comment]
Ok, can somebody else give a somewhat careful review and then we can merge? Maybe you can add some documentation @graingert?

**Comment by graingert** [comment]
@Tarrasch docs added

**Comment by erikbern** [comment]
LGTM

**Comment by Tarrasch** [comment]
Cool, let's see how this does in the wild now :)

**Comment by graingert** [comment]
@Tarrasch do you know when the next release might be?  Is there an official release schedule?

**Comment by Tarrasch** [comment]
No, it'll take a while (month?) from now I think. Unless somebody really asks for it. I'll have to release a 2.0.0 since things have changed.

File: luigi/rpc.py [hunk 10]
```diff
@@ -42,32 +57,62 @@ def __init__(self, message, sub_exception=None):
...
-        self._host = host
-        self._port = port
-        self._url_prefix = url_prefix
+    def __init__(self, url='http://localhost:8082/', connect_timeout=None):
+        assert (
+            not (url.startswith('http+unix://') and not HAS_UNIX_SOCKETS),
+            'You need to install requests-unixsocket for Unix socket support.',
+        )
 
+        self._url = url
```
Comment (paxan) [inline luigi/rpc.py]: If trailing `/` is present on `url`, fetcher gets lots of 404s for path `/app/...`
So it's better to do:

    self._url = url.rstrip('/')

Comment (Tarrasch) [inline luigi/rpc.py]: Would be nice if you could whip up a PR with testcase+fix. :)

Comment (graingert) [inline luigi/rpc.py]: This was possible before with the url_prefix kwargument, so I did this intentionally. But yes the other behaviour makes way more sense

**Comment by geowurster** [comment]
@graingert @Tarrasch @erikbern I think this PR may have broken something involving the worker-master communication, although it could be an issue with my cluster configuration.  For reference, I'm using Google's [`bdutil 1.3.1 (latest)`](https://cloud.google.com/hadoop/bdutil?hl=en) to configure and deploy a hadoop cluster on Google Compute Engine.  After installing the latest commit on master my workers are not able to connect to the scheduler:

```python
ERROR:luigi-interface:Failed connecting to remote scheduler 'http://luigi-master:8082/'
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/luigi/rpc.py", line 124, in _fetch
    response = self._fetcher.fetch(full_url, body, self._connect_timeout)
  File "/usr/local/lib/python2.7/dist-packages/luigi/rpc.py", line 71, in fetch
    raise FetcherException(e)
FetcherException
WARNING: Failed pinging scheduler
WARNING:luigi-interface:Failed pinging scheduler
```

I tried several different commits and I believe fb9a0a2370f47d238e8df9234de8a3f6dbb332e9 is the latest functional commit and 0a3797911425f6f16812ac441f14c4e5e0a2e4c8 introduced the change.

The [blame](https://github.com/spotify/luigi/blame/master/luigi/rpc.py#L71) shows that the object raising this exception was added in a commit in this PR. The annotated screenshot might help to resolve the timeline.

Is anyone else on GCE experiencing this behavior?

commits_ _spotify_luigi

**Comment by graingert** [comment]
@geowurster what's the "original_exc" from the worker?

**Comment by geowurster** [comment]
@graingert `HTTP Error 404: Not Found`.

**Comment by graingert** [comment]
that's odd

File: luigi/rpc.py (hunk 10 above)
Comment (geowurster) [inline luigi/rpc.py]: @graingert I think I'm seeing this problem.  Looking at the `URLLibFetcher()`:

```python
class URLLibFetcher(object):
    def fetch(self, full_url, body, timeout):
        try:
            body = urlencode(body).encode('utf-8')
            return urlopen(full_url, body, timeout).read().decode('utf-8')
        except (URLError, socket.timeout) as e:
            raise FetcherException(e)
```

The `full_url` is `http://luigi-master:8082//api/add_task`, which has a double slash after the port.  Changing that to a single slash seems to fix the problem.

Comment (paxan) [inline luigi/rpc.py]: @Tarrasch here you go:
https://github.com/spotify/luigi/pull/1147

**Comment by Tarrasch** [comment]
@graingert Can you commit to fixing this? I think it's pretty urgent if people can't rely on HEAD anymore! Is #1147 by @paxan related?

**Pull request merged**